Casted unicode object to string to fix subscription_id string error

jayden-at-arista opened the pull request:
Fixes #1322

msftclas commented:
This seems like a small (but important) contribution, so no Contribution License Agreement is required at this point. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

jayden-at-arista commented:
I'm aware of the unicode issue failing the travis build, I'll update the PR once it's fixed.

lmazuel commented:
Hi @jayden-at-arista 
Thanks for the contribution!

May I ask you to add a regression test here:
https://github.com/Azure/azure-sdk-for-python/blob/master/azure-mgmt/tests/test_common.py

Meaning, a test that only succeeds in your code is in.

But the idea of the code LGTM. I might use `six` instead to avoid the python version test. I can't look at it today, but it's high priority for this week for sure.

Thanks!!!

lmazuel commented:
@brettcannon for advice

brettcannon commented on azure-common/azure/common/client_factory.py:
```diff
@@ -25,6 +26,8 @@ def _instantiate_client(client_class, **kwargs):
     if 'subscription_id' not in args:
         del kwargs['subscription_id']
+    elif sys.version_info < (3, 0) and isinstance( kwargs['subscription_id'], unicode ):
+        kwargs['subscription_id'] = str( kwargs['subscription_id'] )
```
If you're going to be casting text down to `str` from `unicode` in Python 2.7 then you should encode with an explicit encoding. There's also unnecessary spaces within the parentheses.

jayden-at-arista commented:
@brettcannon I removed the padding spaces and added explicit encoding. @lmazuel I added regression test, and confirmed it fails without my change and passes with my change. The issue was that the mock clients didn't mock the parameter checking that the real clients perform.

jayden-at-arista commented:
@lmazuel I feel the code is more self documenting with the explicit version check, as this is a band-aid fix of a bigger incompatibility issue. If we wanted to change this to use the six library, we would (from what I saw when experimenting with it) just check to see if it was of type `six.string_types` and then blindly cast it to `str`, which doesn't really show why the cast is necessary.

lmazuel commented:
@jayden-at-arista I figured the mock as well at 2am in the night, waking up "damn, the mock" (true story lol). LGTM. @brettcannon have you still a concern?

brettcannon commented:
@lmazuel nope, single concern was taken care of by @jayden-at-arista 😄

lmazuel commented:
So we're goot to go! Thanks @jayden-at-arista for the PR!!

lmazuel commented:
https://pypi.python.org/pypi/azure-common/1.1.8
